File: Examples/003/CPP/Problem/halo_exchange.cpp
#include <iostream>
#include <vector>
#include <mpi.h>
#include <cmath>

void compute_stencil(int start, int end, const std::vector<double>& data, std::vector<double>& new_data) {
    for (int i = start; i < end; ++i) {
        double val = (data[i - 1] + data[i] + data[i + 1]) / 3.0;
        for (int w = 0; w < 2; ++w) { // Very light math
	  //val = std::sin(val) * std::cos(val) + 1.0;
	  //val = (val * 0.0125) + 0.001;
	  val = std::sin(val)+1.0;
        }
        new_data[i] = val;
    }
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    const int local_elements = 600000; 
    const int halo_size = 500000; // Heavy 4MB Payload
    const int num_steps = 200; 

    std::vector<double> data(local_elements + 2 * halo_size, 1.0);
    std::vector<double> new_data(local_elements + 2 * halo_size, 1.0);

    int left = (rank == 0) ? MPI_PROC_NULL : rank - 1;
    int right = (rank == size - 1) ? MPI_PROC_NULL : rank + 1;

    MPI_Barrier(MPI_COMM_WORLD);
    double start_time = MPI_Wtime();

    for (int step = 0; step < num_steps; ++step) {
        // Sync communication: CPU idles during the 4MB transfer
        MPI_Sendrecv(&data[local_elements], halo_size, MPI_DOUBLE, right, 0,
                     &data[0],              halo_size, MPI_DOUBLE, left,  0,
                     MPI_COMM_WORLD, MPI_STATUS_IGNORE);

        MPI_Sendrecv(&data[halo_size],                       halo_size, MPI_DOUBLE, left,  1,
                     &data[halo_size + local_elements],      halo_size, MPI_DOUBLE, right, 1,
                     MPI_COMM_WORLD, MPI_STATUS_IGNORE);

        compute_stencil(halo_size, halo_size + local_elements, data, new_data);
        data.swap(new_data);
    }

    double max_t, total = MPI_Wtime() - start_time;
    MPI_Reduce(&total, &max_t, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
    if (rank == 0) std::cout << "SYNC Heavy-Net Time: " << max_t << "s" << std::endl;

    MPI_Finalize();
    return 0;
}

File: Examples/004/Fort/Problem/halo_exchange.f90
program halo_exchange
    use mpi
    implicit none

    integer :: ierr, rank, size, step, i, w
    integer, parameter :: local_elements = 600000
    integer, parameter :: halo_size = 500000
    integer, parameter :: num_steps = 200
    integer :: left, right
    real(8), dimension(:), allocatable :: data_arr, new_data
    real(8) :: start_time, total_time, max_time
    integer :: total_size

    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)

    total_size = local_elements + 2 * halo_size
    allocate(data_arr(total_size))
    allocate(new_data(total_size))
    data_arr = 1.0d0
    new_data = 1.0d0

    left = rank - 1
    right = rank + 1
    if (rank == 0) left = MPI_PROC_NULL
    if (rank == size - 1) right = MPI_PROC_NULL

    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    start_time = MPI_Wtime()

    do step = 1, num_steps
        ! Sync communication
        call MPI_Sendrecv(data_arr(halo_size + 1), halo_size, MPI_DOUBLE_PRECISION, left, 1, &
                          data_arr(halo_size + local_elements + 1), halo_size, MPI_DOUBLE_PRECISION, right, 1, &
                          MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

        call MPI_Sendrecv(data_arr(local_elements + 1), halo_size, MPI_DOUBLE_PRECISION, right, 0, &
                          data_arr(1), halo_size, MPI_DOUBLE_PRECISION, left, 0, &
                          MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

        ! Encapsulated stencil call
        call compute_stencil(halo_size + 1, halo_size + local_elements, data_arr, new_data)
        
        data_arr = new_data
    end do

    total_time = MPI_Wtime() - start_time
    call MPI_Reduce(total_time, max_time, 1, MPI_DOUBLE_PRECISION, MPI_MAX, 0, MPI_COMM_WORLD, ierr)

    if (rank == 0) print *, "SYNC Heavy-Net Time Fortran: ", max_time, "s"

    deallocate(data_arr, new_data)
    call MPI_Finalize(ierr)

contains

    subroutine compute_stencil(start_idx, end_idx, d, nd)
        integer, intent(in) :: start_idx, end_idx
        real(8), dimension(:), intent(in) :: d
        real(8), dimension(:), intent(out) :: nd
        integer :: j, k
        real(8) :: val

        do j = start_idx, end_idx
            val = (d(j-1) + d(j) + d(j+1)) / 3.0d0
            do k = 1, 2
                val = sin(val) + 1.0d0
            end do
            nd(j) = val
        end do
    end subroutine compute_stencil

end program halo_exchange
